### PR TITLE
Allow to pass initial data to the editor constructor

### LINKED ISSUE #72: Allow to set data in constructor instead of element
There should be a possibility to initialize editor with data passed to the constructor instead of the element to replace. Part of https://github.com/ckeditor/ckeditor5-core/issues/64. After initializing the editor `editor.element` should be added manually to the DOM.

### PR DESCRIPTION
**Pull request by szymonkups**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Editor can be created with initial data passed to the constructor. Closes #72.

BREAKING CHANGE: `ClassicEditor#element` is now available as `ClassicEditor#sourceElement`. See ckeditor/ckeditor5-core#64.

Requires: https://github.com/ckeditor/ckeditor5-core/pull/129

### CODE REVIEW

**Comment by Reinmar** [comment]
R- due to https://github.com/ckeditor/ckeditor5-core/issues/64#issuecomment-395069721.

File: src/classiceditor.js [hunk 1]
```diff
@@ -156,8 +155,22 @@ export default class ClassicEditor extends Editor {
 	 *				console.error( err.stack );
 	 *			} );
 	 *
-	 * @param {HTMLElement} element The DOM element that will be the source for the created editor.
-	 * The data will be loaded from it and loaded back to it once the editor is destroyed.
+	 * @param {HTMLElement|String} elementOrData The DOM element that will be the source for the created editor
+	 * or editor's initial data.
+	 *
+	 * If an element is passed, then it contents will be automatically
```
Comment (Reinmar) [inline src/classiceditor.js]: its contents

File: src/classiceditor.js [hunk 2]
```diff
@@ -156,8 +155,22 @@ export default class ClassicEditor extends Editor {
...
+	 * {@link module:editor-classic/classiceditor~ClassicEditor#setData loaded} to the editor on startup
+	 * and  the editor element will replace the passed element in the DOM (the original one will be hidden and editor
```
Comment (Reinmar) [inline src/classiceditor.js]: "editor element" -> link to the new property

File: src/classiceditor.js [hunk 3]
```diff
@@ -156,8 +155,22 @@ export default class ClassicEditor extends Editor {
...
+	 * will be injected next to it).
+	 *
+	 * Moreover, the data will be set back to the original element once the editor is destroyed and
```
Comment (Reinmar) [inline src/classiceditor.js]: original => source

File: src/classiceditor.js [hunk 4]
```diff
@@ -79,15 +83,24 @@ export default class ClassicEditor extends Editor {
 		attachToForm( this );
 	}
 
+	/**
+	 * An HTML element that represents the whole editor (with UI). See the {@link module:core/editor/editorwithui~EditorWithUI} interface.
```
Comment (Reinmar) [inline src/classiceditor.js]: Do we have to redefine this API doc at all?

File: src/classiceditor.js [hunk 5]
```diff
@@ -128,25 +141,72 @@ export default class ClassicEditor extends Editor {
...
-						editor._elementReplacer.replace( element, editor.ui.view.element );
+						if ( isElement( sourceElementOrData ) ) {
+							editor._elementReplacer.replace( sourceElementOrData, editor.ui.view.element );
+						}
+
 						editor.fire( 'uiReady' );
 					} )
 					.then( () => editor.editing.view.attachDomRoot( editor.ui.view.editableElement ) )
-					.then( () => editor.data.init( getDataFromElement( element ) ) )
+					.then( () => {
```
Comment (Reinmar) [inline src/classiceditor.js]: It'd be easier to read if it was written like this:

```js
const data = isElement(  sourceElementOrData ) ? getDataFromElement( sourceElementOrData ) : sourceElementOrData;

editor.data.init( data );
```

Comment (Reinmar) [reply src/classiceditor.js]: The point is – what differs these two scenarios isn't how or whether you call `editor.data.init()`. It's how you retrieve these data.

File: src/classiceditor.js [hunk 6]
```diff
@@ -128,25 +141,72 @@ export default class ClassicEditor extends Editor {
...
+						if ( editor.sourceElement ) {
+							editor.data.init( getDataFromElement( editor.sourceElement ) );
+						} else {
+							editor.data.init( sourceElementOrData );
+							editor.sourceElement = editor.ui.view.element;
```
Comment (Reinmar) [inline src/classiceditor.js]: I think it's a mistake. Probably a leftover.

Comment (Reinmar) [reply src/classiceditor.js]: BTW, how did tests pass with this? There must be a missing test.

**Comment by pomek** [comment]
@Reinmar, would you like to check it once again?

File: src/classiceditor.js (hunk 5 above)
Comment (Reinmar) [reply src/classiceditor.js]: Note that the code you committed isn't equal to what I proposed:

```js
editor.data.init(
    isElement( sourceElementOrData ) ? getDataFromElement( sourceElementOrData ) : sourceElementOrData
);
```

In this case, we lose the information what do we call `init()` with. It makes reading the code harder.

Comment (Reinmar) [reply src/classiceditor.js]: BTW, `editor.data.init()` may return a promise so this change breaks the promise chain – make sure to add a test for that.

**Comment by Reinmar** [comment]
Here you are :D

**Comment by pomek** [comment]
Fixed.

**Comment by Reinmar** [comment]
I'll be pedantic – it's still not the same as what I proposed. I named the variable `initialData` on purpose to make it clear that it's not just any data, but the data you load initially. Compared to a simple `data` it brings more information on what's happening. Just like, not every `key` in an object is just a key. They are command names, element names, plugin names, keystrokes, whatever. 

It's not critical here. It's not even a place when one may have some bigger doubts. But did you think about this when changing the name? Cause that should've been intentional and I don't think it was.

**Comment by pomek** [comment]
When the name is `initialData`, length of the line is bigger than 140 and it requires additional eslint rules... That's the reason why I'm trying "smuggle" other name.

**Comment by Reinmar** [comment]
That is the worst possible reason ;) Try formatting that piece of code differently (split into multiple lines if needed).

**Pull request merged**